libs/mrwebrtc-unityplugin/src/native_renderer.cpp

```diff
@@ -133,45 +167,74 @@ void NativeRenderer::DisableRemoteVideo() {
-    // preferring to copy the frame buffer outside any lock. Alternatively, the
-    // copy operation could be done inside g_lock scope. This would result in a
-    // single g_lock, but the lock would be held for a longer period of time
-    // every video frame, for every video stream. RESEARCH: Which is better?
+  NativeRenderer* nativeVideo = static_cast<NativeRenderer*>(user_data);
+ 
+  {
+    // Instance lock in case they get updated from UpdateTextures
+    std::lock_guard guard(nativeVideo->m_lock);
+    if ((int)frame.width_ != nativeVideo->m_remoteTextures[0].width || (int)frame.height_ != nativeVideo->m_remoteTextures[0].height) {
+      if (g_textureSizeChangeCallback != nullptr) {
+        g_textureSizeChangeCallback(frame.width_, frame.height_, nativeVideo->m_handle);
+      }
+      return;
+    }
+  }
 
-    // Acquire a video frame buffer from the free list.
-    {
-        // Global lock
-        std::lock_guard guard(g_lock);
-        if (nativeVideo->m_nextI420RemoteVideoFrame == nullptr) {
-            // TODO: Encapsulate this logic.
-            if (g_freeI420VideoFrames.size()) {
-                nativeVideo->m_nextI420RemoteVideoFrame = g_freeI420VideoFrames.back();
-                g_freeI420VideoFrames.pop_back();
-            } else {
-                nativeVideo->m_nextI420RemoteVideoFrame = std::make_shared<I420VideoFrame>();
-            }
-        }
+  // Acquire a video frame buffer from the free list with the global lock.
+  std::shared_ptr<I420VideoFrame> remoteI420Frame = nullptr;
+  {
+    // Global lock
+    std::lock_guard guard(g_lock);
+    if (g_freeI420VideoFrames.size()) {
+        remoteI420Frame = std::move(g_freeI420VideoFrames.back());
+        g_freeI420VideoFrames.pop_back();
+    } else {
+      try {
+        remoteI420Frame = std::make_shared<I420VideoFrame>();
+      } catch (std::bad_alloc const&) {
+        return;
+      }
     }
+  }
 
-    // Copy the incoming video frame to the buffer.
-    nativeVideo->m_nextI420RemoteVideoFrame->CopyFrame(frame);
 
-    // Queue the renderer for the next video update.
-    {
-        // Global lock
-        std::lock_guard guard(g_lock);
-        // Register this renderer for the next video update.
-        g_nativeVideoUpdateQueue.emplace(nativeVideo);
+  // Copy frame outside of any lock for perf.
+  // This may consume more time on this thread unnecessarily, but
+  // will block the render update thread for less time.
+  bool frameCopySuccess = remoteI420Frame->CopyFrame(frame);
+
+  bool shouldUpdateFrame = false;
+  if (frameCopySuccess) {
+    // Set new copied frame on nativeVideo while in the instance lock of the
+    // NativeVideo otherwise the render update loop may grab it pre-emptively.
+    std::lock_guard guard(nativeVideo->m_lock);
+    // Only update if null, otherwise this means the frame previously
+    // added has not yet been processed.
+    if (nativeVideo->m_nextI420RemoteVideoFrame == nullptr) {
```

Comment:
Are we saying here that we drop the current frame in favor of keeping an old frame? Shouldn't we do the opposite?

Reply:
Yes, that is probably more correct.